### ISSUE TITLE
RiakCS 1.5.2 shows sibling explosion while doing rapid updates during join/leave operations

### ISSUE THREAD

**Issue by bsparrow435** [basho/riak_cs#1005]
Observed at a customer site as well as locally in testing. Example test:
- Create 2 node, ring_size=4 Riak/RiakCS cluster with default backend and other settings
- Load 1000 2MB objects into RiakCS
- Spawn 10 threads each rapidly updating the same object in a for loop:

```
for i in {1..200}; do ./s3cmd -c ~/s3cfg put test_file s3://test-bucket/; done
```
- Leave, plan, commit node 2 from the cluster.
- Observe sibling growth on node 1 in the logs(This is easier if you set the threshold low: `{warn_siblings,2}` in riak_kv of app.config.

-Brian

**Comment by shino** [comment]
Thank you for tracking down the issue.

It seems that maybe_backpressure_sleep does harm.
It was added mainly for multipart upload because PUT Part
 can be (massively) concurrent.

At normal (non-multipart) put, sleep after `add_new_dict` is bad idea.
Or (better?) we should set #state.riak_object/manifest to undefined
if we execute backpressure sleep in order not to use the previous result..

One question. PUT Object in concurrent way seems not usual use cases
because CS is just last-write-wins. Why concurrent uploads in the customer?
Simulating (very) rapid retries?

**Comment by shino** [comment]
Relation with cluster change (leave) is not understood yet. Just an indirect factor to cause an additional load to the cluster or relevant directly in some way.

**Comment by shino** [comment]
The change included in 1.5.1 https://github.com/basho/riak_cs/pull/959

**Comment by bsparrow435** [comment]
The customer updates a small subset of objects once every 2 seconds and then a larger subset every 10 seconds. When things get slow, these updates can be concurrent because they may take longer than 2 and 10 seconds respectively.

The explosion has only been observed during the cluster leave operation. When transfers were disabled, vnode queues began to drop on the vnodes with high siblings and the siblings are no longer generated.

**Comment by kuenishi** [comment]
@shino @bsparrow435 do you mean this is the case because this does not happen in CS 1.4? Uploads with high concurrency like > 10 might cause this regardless of CS version.

**Comment by shino** [comment]
> Uploads with high concurrency like > 10 might cause like this
>   regardless of CS version, to me...

Agree. This causes interleaving get-modify-put accesses, then any version
of CS with Riak 1.4.x series can suffer from sibling explosion.

One thing we already know is, with CS 1.5, things go worse than 1.4
when siblings > 5 by backpressure sleep. But, up to siblings <= 5, I
don't know there is a difference or not.

Are 5 siblings generated only by two-second-interval uploads?
Is there another factor? Hmm...

**Comment by kuenishi** [comment]
> One thing we already know is, with CS 1.5, things go worse than 1.4
> when siblings > 5 by backpressure sleep. But, up to siblings <= 5, I
> don't know there is a difference or not.

I will try some numerical measurement and report back later.

**Comment by shino** [comment]
Siblings can be caused by [1].
During handoff, siblings grow in target (receiving) vnodes. Once handoff finishes, the target vnodes
becomes a new owner, then client (GET FSM or PUT FSM) see possibly many siblings.
Then backpressure sleep added in 1.5.1 [2] will be triggered in CS nodes.

If PUT Object requests are interleaved because of backpressure sleep,
higher possibility of creating siblings further. This bug is filed at [3].

[1] https://github.com/basho/riak_kv/issues/1046
[2] https://github.com/basho/riak_cs/pull/959
[3] https://github.com/basho/riak_cs/issues/1022

**Comment by kuenishi** [comment]
Closing as 1.5.3 has been released and remaining task will be tracked at #1022.
